File: src/power_management.cpp
// Implémentation gestion alimentation

#include "power_management.hpp"
#include "driver/i2c.h"
#include <cstdio>
#include <cmath>

// I2C pins pour ESP32
#define I2C_MASTER_NUM I2C_NUM_0
#define I2C_MASTER_SDA_IO 21
#define I2C_MASTER_SCL_IO 22
#define I2C_MASTER_FREQ_HZ 100000

Battery::Battery(uint8_t id, uint8_t i2c_addr)
    : battery_id_(id), i2c_addr_(i2c_addr), voltage_(0), current_(0),
      power_(0), is_connected_(false), max_voltage_(12), min_voltage_(9) {

    printf("[Battery%d] Initialized at I2C address 0x%02X\n", id, i2c_addr);
}

void Battery::read_voltage() {
    // Simulation pour tests (remplacer par vrai lecteur INA219)
    voltage_ = max_voltage_ - (random() % 30) / 10.0f;
    is_connected_ = true;

    printf("[Battery%d] Voltage: %.2fV\n", battery_id_, voltage_);
}

void Battery::read_current() {
    // Simulation pour tests
    current_ = (random() % 50) * 10.0f;  // 0-500mA

    printf("[Battery%d] Current: %.1fmA\n", battery_id_, current_);
}

void Battery::read_power() {
    read_voltage();
    read_current();

    power_ = (voltage_ * current_) / 1000.0f;

    printf("[Battery%d] Power: %.2fW\n", battery_id_, power_);
}

void Battery::check_low_voltage() {
    if (voltage_ < min_voltage_) {
        printf("[Battery%d] WARNING: Low voltage (%.2fV < %.2fV)\n",
               battery_id_, voltage_, min_voltage_);
    }
}

BatteryStatus Battery::get_status() const {
    return BatteryStatus{
        .battery_id = battery_id_,
        .voltage = voltage_,
        .current = current_,
        .power = power_,
        .is_connected = is_connected_
    };
}

// ============= PowerManagement =============

PowerManagement::PowerManagement()
    : battery1(1, 0x40), battery2(2, 0x41),
      voltage_12v_(12), voltage_5v_(5), voltage_3v3_(3.3),
      emergency_stop_active_(false) {

    // Initialiser I2C
    i2c_config_t conf = {
        .mode = I2C_MODE_MASTER,
        .sda_io_num = I2C_MASTER_SDA_IO,
        .scl_io_num = I2C_MASTER_SCL_IO,
        .master = {.clk_speed = I2C_MASTER_FREQ_HZ},
    };

    i2c_param_config(I2C_MASTER_NUM, &conf);
    i2c_driver_install(I2C_MASTER_NUM, I2C_MODE_MASTER, 0, 0, 0);

    printf("[PowerManagement] I2C initialized\n");
}

void PowerManagement::read_all_voltages() {
    battery1.read_voltage();
    battery2.read_voltage();

    printf("[PowerManagement] Total voltage: %.2fV\n",
           battery1.voltage_ + battery2.voltage_);
}

void PowerManagement::read_all_currents() {
    battery1.read_current();
    battery2.read_current();

    printf("[PowerManagement] Total current: %.1fmA\n",
           battery1.current_ + battery2.current_);
}

bool PowerManagement::check_system_health() {
    read_all_voltages();

    bool bat1_ok = battery1.voltage_ >= battery1.min_voltage_;
    bool bat2_ok = battery2.voltage_ >= battery2.min_voltage_;

    if (!bat1_ok || !bat2_ok) {
        printf("[PowerManagement] WARNING: Low voltage detected!\n");
        return false;
    }

    printf("[PowerManagement] System healthy\n");
    return true;
}

SystemPowerStatus PowerManagement::get_summary() {
    read_all_voltages();
    read_all_currents();
    battery1.read_power();
    battery2.read_power();

    return SystemPowerStatus{
        .battery1 = battery1.get_status(),
        .battery2 = battery2.get_status(),
        .system_voltage = battery1.voltage_ + battery2.voltage_,
        .system_current = battery1.current_ + battery2.current_,
        .system_power = battery1.power_ + battery2.power_,
        .emergency_stop_active = emergency_stop_active_
    };
}

void PowerManagement::emergency_stop() {
    emergency_stop_active_ = true;
    printf("[PowerManagement] EMERGENCY STOP ACTIVATED\n");
}

void PowerManagement::resume_operation() {
    emergency_stop_active_ = false;
    printf("[PowerManagement] Resumed operation\n");
}

float PowerManagement::estimate_runtime(float consumption_per_hour_wh) {
    float total_voltage = battery1.voltage_ + battery2.voltage_;
    float total_capacity_wh = total_voltage * 10;  // Estimation simple

    float runtime_hours = total_capacity_wh / consumption_per_hour_wh;

    printf("[PowerManagement] Estimated runtime: %.1f hours\n", runtime_hours);
    return runtime_hours;
}

File: power-management.js
// Gestion de l'alimentation et mesure de tension (INA219)

const { PINS, POWER } = require('./esp32-pins');

class Battery {
  constructor(batteryId = 1, ina219Address = null) {
    this.batteryId = batteryId;
    this.ina219Address = ina219Address || [PINS.INA219_BATTERY_1, PINS.INA219_BATTERY_2][batteryId - 1];
    this.voltage = 0; // V
    this.current = 0; // mA
    this.power = 0; // W
    this.isConnected = false;
    this.calibrationFactor = 0.1; // Pour convertisseur 10x (exemple)
    this.maxVoltage = 12; // ou 24V selon config
    this.minVoltage = 9; // seuil alerte basse tension
    this.history = [];
  }

  // Lire la tension via I2C INA219
  readVoltage() {
    // Code réel ESP32:
    // const raw = ina219.readBusVoltage();
    // this.voltage = raw / 1000; // en V

    // Simulation pour tests:
    this.voltage = this.maxVoltage - (Math.random() * 0.5);
    this.isConnected = true;

    const timestamp = new Date().toISOString();
    this.history.push({
      timestamp,
      voltage: this.voltage,
      current: this.current,
      power: this.power,
    });

    return this.voltage;
  }

  // Lire le courant via I2C INA219
  readCurrent() {
    // Code réel ESP32:
    // this.current = ina219.readCurrent_mA();

    // Simulation:
    this.current = Math.random() * 500; // 0-500mA

    return this.current;
  }

  // Lire la puissance
  readPower() {
    this.readVoltage();
    this.readCurrent();
    this.power = (this.voltage * this.current) / 1000; // en W
    return this.power;
  }

  // Vérifier la santé de la batterie
  getHealth() {
    const percent = (this.voltage / this.maxVoltage) * 100;
    const status = percent > 80 ? '🟢 Excellente' :
                   percent > 60 ? '🟡 Bonne' :
                   percent > 40 ? '🟠 Faible' :
                   '🔴 Critique';

    return {
      percent: percent.toFixed(1),
      status,
      voltage: this.voltage.toFixed(2),
    };
  }

  // Avertissement si tension basse
  checkLowVoltage() {
    if (this.voltage < this.minVoltage) {
      console.log(`⚠️ Batterie ${this.batteryId}: TENSION BASSE (${this.voltage.toFixed(2)}V)`);
      return true;
    }
    return false;
  }

  // Obtenir l'état
  getStatus() {
    return {
      batteryId: this.batteryId,
      ina219Address: `0x${this.ina219Address.toString(16)}`,
      isConnected: this.isConnected,
      voltage: this.voltage.toFixed(2),
      current: this.current.toFixed(1),
      power: this.power.toFixed(2),
      health: this.getHealth(),
      maxVoltage: this.maxVoltage,
      minVoltage: this.minVoltage,
    };
  }

  printStatus() {
    const s = this.getStatus();
    console.log(`\n🔋 État Batterie ${s.batteryId}:`);
    console.log(`   I2C: ${s.ina219Address} (${s.isConnected ? '🟢 Connectée' : '🔴 Déconnectée'})`);
    console.log(`   Tension: ${s.voltage}V (max: ${s.maxVoltage}V, min: ${s.minVoltage}V)`);
    console.log(`   Courant: ${s.current}mA`);
    console.log(`   Puissance: ${s.power}W`);
    console.log(`   Santé: ${s.health.status} (${s.health.percent}%)`);
    return this;
  }
}

// Classe pour gérer les deux batteries
class PowerManagement {
  constructor() {
    this.battery1 = new Battery(1, PINS.INA219_BATTERY_1);
    this.battery2 = new Battery(2, PINS.INA219_BATTERY_2);
    this.voltage12V = 12;
    this.voltage5V = 5;
    this.voltage3V3 = 3.3;
    this.emergencyStopActive = false;
  }

  // Lire toutes les tensions
  readAllVoltages() {
    this.battery1.readVoltage();
    this.battery2.readVoltage();

    return {
      battery1: this.battery1.voltage,
      battery2: this.battery2.voltage,
      total: this.battery1.voltage + this.battery2.voltage,
    };
  }

  // Lire tous les courants
  readAllCurrents() {
    this.battery1.readCurrent();
    this.battery2.readCurrent();

    return {
      battery1: this.battery1.current,
      battery2: this.battery2.current,
      total: this.battery1.current + this.battery2.current,
    };
  }

  // Vérifier la santé générale
  checkSystemHealth() {
    const bat1Low = this.battery1.checkLowVoltage();
    const bat2Low = this.battery2.checkLowVoltage();

    if (bat1Low || bat2Low) {
      console.log('⚠️ ⚠️ ALERTE: Tension batterie basse');
      return false;
    }

    console.log('✅ Système électrique sain');
    return true;
  }

  // Obtenir un résumé du système
  getSummary() {
    return {
      battery1: this.battery1.getStatus(),
      battery2: this.battery2.getStatus(),
      systemVoltage: (this.battery1.voltage + this.battery2.voltage).toFixed(2),
      systemCurrent: (this.battery1.current + this.battery2.current).toFixed(1),
      systemPower: (this.battery1.power + this.battery2.power).toFixed(2),
      emergencyStop: this.emergencyStopActive,
    };
  }

  // Afficher un résumé complet
  printSummary() {
    const s = this.getSummary();
    console.log('\n⚡ RÉSUMÉ SYSTÈME ÉLECTRIQUE:');
    console.log(`   Tension totale: ${s.systemVoltage}V`);
    console.log(`   Courant total: ${s.systemCurrent}mA`);
    console.log(`   Puissance totale: ${s.systemPower}W`);
    console.log(`   Arrêt d'urgence: ${s.emergencyStop ? '🔴 ACTIVÉ' : '🟢 Inactif'}`);
    this.battery1.printStatus();
    this.battery2.printStatus();
    return this;
  }

  // Activer arrêt d'urgence
  emergencyStop() {
    this.emergencyStopActive = true;
    console.log('🛑 🛑 ARRÊT D\'URGENCE ACTIVÉ 🛑 🛑');
    // Couper toutes les alimentations
    return this;
  }

  // Désactiver arrêt d'urgence
  resumeOperation() {
    this.emergencyStopActive = false;
    console.log('✅ Reprise d\'opération après arrêt d\'urgence');
    return this;
  }

  // Obtenir le rapport d'autonomie estimée
  estimateRuntime(consumptionPerHourWh = 100) {
    const total_Wh = (this.battery1.voltage + this.battery2.voltage) * 10; // Estimation simple
    const runtimeHours = total_Wh / consumptionPerHourWh;
    const runtimeMinutes = runtimeHours * 60;

    return {
      estimatedRuntimeHours: runtimeHours.toFixed(1),
      estimatedRuntimeMinutes: runtimeMinutes.toFixed(0),
      consumptionPerHour: consumptionPerHourWh,
    };
  }

  printRuntime(consumptionPerHourWh = 100) {
    const rt = this.estimateRuntime(consumptionPerHourWh);
    console.log(`\n⏱️ AUTONOMIE ESTIMÉE:`);
    console.log(`   Consommation: ${rt.consumptionPerHour}Wh/h`);
    console.log(`   Durée: ~${rt.estimatedRuntimeHours}h (${rt.estimatedRuntimeMinutes}min)`);
    return this;
  }
}

module.exports = { Battery, PowerManagement };
